manpages/raw2mrc.man
.na
.nh
.TH raw2mrc 1 2.50 BL3DEMC
.SH NAME
raw2mrc \- Create an MRC image stack from raw image data.
.SH SYNOPSIS
raw2mrc  [options]  raw_image_files...  MRC_output_file
.SH DESCRIPTION
raw2mrc can be used to create MRC image files from many types
of raw binary image data.  
Command line options are used to adjust the image size and
to skip the input image header.
Several raw images can be combined into one file by listing all of their
filenames before the name of the output file.
.SH OPTIONS
.TP
.B -x \fIwidth\fR
Set the width of the image, the default value is 640 pixels.
.TP
.B -y \fIheight\fR
Set the height of the input image, 
the default is 480 pixels.
.TP
.B -z \fIsections\fR
Set the number of sections in each input image file,
the default is 1.
.TP
.B -t \fItype\fR
Set the type of input data, the default byte data. 
The allowed types are
byte, short, ushort, long, ulong, float and rgb.
Where byte is 8-bit integer values, short is 16-bit signed values,
ushort is 16-bit unsigned values, long is 32-bit signed values,
ulong is 32-bit unsigned values, float is a 32-bit floating 
point value and rgb is three byte values--one for each color.
.TP
.B -s
Swap input bytes.
.TP
.B -o \fIoffset\fR
Set the offset value used at the beginning of the file, 
the default is 0.
.TP
.B -oz \fIoffset\fR
Set the offset value used at the end of each section,
the default offset is 0.
.TP
.B -f
Flip the images around the X axis.
.TP
.B -i
Produce an inverted output stack by both inverting the order in which files
are processed and inverting the order of reading Z slices from each file.
.TP
.B -iz
Invert the order of reading Z slices from each individual file.
.TP
.B -d
Divide unsigned 16-bit values or any 32-bit values by 2.  The default 
for unsigned short values is to subtract 32767, which
allows all unsigned short values to be stored as signed short integers in the
MRC file of mode 1.  Use this option if you do not want negative numbers and
do not care about the loss of precision.  If the values in the input files are
all
less than 32768, then specify that the input data are short rather than ushort.
.TP
.B -u
Store unsigned 16-bit values without modification in an MRC file of mode 6,
which is a non-standard mode that may not be completely supported.
.TP
.B -c
Convert 32-bit integers to 16-bit integers instead of to floating point
numbers.  If the type is long, you can use this option alone if values are 
between -32768 and 32767; or you can use the option with
.B -d
to divide by 2 if values are between -65536 and 65535.  If you specify the
type as ulong, then by default 32767 will be subtracted from the input values,
so that values from 0 to 65534 can be stored as -32767 to 32767.  Again,
.B -d
can be used to divide by 2 instead, so that values from 0 to 65534 can be
stored as 0 to 32767.  Also, 
.B -u
can be used to store the numbers as unsigned integers in a file of mode 6.
If values in the input files are all less than 32768, 
then specify the type as long instead of ulong.
.TP
.B -p \fIpixel\fR
Set the pixel spacing of the MRC file header to the given value, which should
generally be a pixel size in Angstroms.
.B -pz \fIpixel\fR
Set the Z pixel spacing of the MRC file header to the given value if it is 
different from the one given with the 
.B -p
option.
.SH FILES
If the output file already exists, it will be made a backup file by adding ~
to its name.
.SH AUTHORS
Jim Kremer, David Mastronarde
.SH SEE ALSO
tif2mrc(1), tomrc(1), frommrc(1), mrcv(1)
.SH BUGS
Email bug reports to mast@colorado.edu.


History
IMOD History
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
12/6/13: Added option to Imodinfo to get contour lengths broken out by
fine-grained color, and made -o option provide object list for all outputs.

12/5/13: Updated the eTomo manpage.  Improved the eTomo help statement.
(Bug# 1764)

12/3/13: In eTomo - PEET, excluding the quoted values functionality.
(Bug# 1631)

11/30/13: Changes to Tiltalign to deal with patch tracking data: it can select
a subset of patches for the global alignment to avoid solving for too many
variables; it can do robust fitting with weights assigned either to individual
points or to whole contours.

11/28/13: Made changes in the search routine used by Tiltalign/Beadtrack to
reduce the high risk of "minimization errors" when starting close to the final
solution; also ignore particular errors after first round of robust fitting.

11/28/13: In eTomo allowing quoted values in autodocs.(Bug# 1631)

11/21/13: IMOD_4-6-33
11/21/13: Changed Corrsearch3d to use dynamic allocation, load from file more 
efficiently, and use multiple threads for some steps.

11/20/13: In eTomo - Final Aligned Stack, prevented the creation of a bad
rotation.xf file when the final aligned image stack is not present.(Bug# 1762)

11/20/13: In eTomo - Serial Sections, fixed the Newstack progress bar.
(Bug# 1749)

11/20/13: Fixed 3dmod crash from loading many files through startup window.

11/20/13: Made multi-line INFO and WARNING messages more parsable in
etomo_err.log.(Bug# 1694)

11/19/13: In eTomo - Combine, prevented the dataset from being damaged when
the Combine dialog is opened too early.(Bug# 1652)

11/19/13: In uitest, fixed a problem with finding the main panel.(Bug# 1761)

11/18/13: In eTomo made multi-line error messages parsable by adding an empty
line after each one.(Bug# 1694)

11/18/13: In eTomo - Coarse Alignment, added support for tiltxcorr parameters
SearchMagChanges and ViewsWithMagChanges.(Bug# 1760)

11/15/13: In eTomo added support for parallel processing using multiple GPUs
per computer.(Bug# 1436)

11/15/13: In eTomo reduced the excessive debug output.(Bug# 1763)

11/13/13: In eTomo - Serial Sections, added a Sobel filter checkbox and
improved the behavior of the alignment panel.(Bug# 1755)

11/11/13: IMOD_4-6-32
11/5/13: Fixed bug that crashed Imodfindbeads when adjusting size downward.

11/4/13: In eTomo - Fiduacial Model dialog, fixed a problem with deleting
autofidseed temporary files in Windows.(Bug# 1759)

11/4/13: Make xfalign apply shift limits in initial cross-correlation.

11/4/13: Fixed scale bar drawing from montage snapshot of model view window.

11/4/13: Made mrc2tif use 4 or 5 digits for all output file numbers if needed.

11/2/13: Switched display of residual model to show arrows.

11/1/13: In eTomo - Serial Sections, improved the context sensitive pop-up
menus.(Bug# 1748)

11/1/13: In eTomo - Fiducial Model, added fields for AdjustSizes and
ClusteredPointsAllowed to the Autofidseed panel.(Bug# 1743)

11/1/13: Made Tiltxcorr able to find mag changes at selected views; for a 
bidirectional series, Copytomocoms puts the option with the view number in
xcorr.com.

11/1/13: Fixed bug in imodtrans -3 option introduced 12/8/12.

10/30/13: IMOD_4-6-31
10/29/13: Fixed a problem in eTomo where the parallel panel did not appear
when eTomo reconnected to a parallel process.  This problem appeared when
Pause and Resume had been previously used.(Bug# 1757)

10/29/13: Fixed the eTomo version string.(Bug# 1744)

10/28/13: In eTomo fixed a bug where stack header data was not getting into
the dataset file when eTomo was run from batchruntomo.(Bug# 1758)

10/27/13: Added option to Tilt to get output in parallel (X/Y) slices without
inverting handedness. 

10/25/13: In eTomo removed dependencies on Cygwin.(Bug# 1713)

10/25/13: Changes in Tiltxcorr: shift limits are now applied in an elliptical
rather than rectangular region; implemented a new, more selective method for
excluding the central peak due to fixed pattern noise; added an option to
compute normalized correlation coefficients at multiple peak positions and
pick the highest (not useful so far).

10/25/13: Changed Beadtrack's elongation file output so that automatic seed
picking would not fail with only 3 beads.

10/24/13: In eTomo - Combine, removed the message complaining about tomograms
A and B not being the same size in X and Y, as this is not an error.
(Bug# 1651)

10/24/13: In eTomo, preventing embedded spaces in dataset names.(Bug# 1667)

10/24/13: Added a antialiasFilter field for the newstack panels in eTomo -
Coarse Alignment and Final Aligned Stack.(Bug# 1737)

10/23/13: In eTomo resizing when the Cleanup dialog is finished.(Bug# 1636)

10/22/13: Fixed another warping-related crash in Midas.

10/22/13: Made it possible to start Zap rubber band from ANY corner.

10/21/13: Fixed a problem that crashed 3dmod when a slicer angle is very
small, such as when doing mouse-rotating.

10/18/13: Added toolbar buttons to Zap and Slicer for drawing a single thick
arrow for snapshots; also added an object property to draw arrowheads at the
ends of contours.

10/17/13: In eTomo - Clean up, added unused, optional image files (such as the
output from mtffilter) to the file chooser.(Bug# 1509)

10/17/13: In eTomo - Clean up, added SIRT .vsr## output files.(Bug# 1533)

10/15/13: Fixed numerical problems in drawing code that crashed Midas on big
images in 64-bit Windows.

10/14/13: Added GiantCriterion, BigDiffCriterion, and ExtraLargeRadius to
eTomo - Preprocessing.(Bug# 1736)

10/9/13: Fixed a problem with running tomodataplots from eTomo under Java 1.7
that prevented the graphs from coming up.(Bug# 1754)

10/9/13: In eTomo fixed a parallel processing problem which prevented the
Resume button from working when the process was pausing.  This problem
appeared in only in reconnected processes.(Bug# 1729)

10/9/13: In eTomo - Nonlinear Anisotropic Diffusion, some test.K... file names

mrc/raw2mrc.c
/*
 *  raw2mrc.c -- Convert raw image data to an mrc image file.
 *
 *  Original author: James Kremer
 *  Revised by: David Mastronarde   email: mast@colorado.edu
 *
 *  Copyright (C) 1995-2005 by Boulder Laboratory for 3-Dimensional Electron
 *  Microscopy of Cells ("BL3DEMC") and the Regents of the University of 
 *  Colorado.  See dist/COPYRIGHT for full copyright notice.
 *
 *  $Id$
 */

#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include "mrcfiles.h"
#include "b3dutil.h"
#include "parse_params.h"

/* input data types */
#define DTYPE_CHAR   1
#define DTYPE_BYTE   1
#define DTYPE_UCHAR  2
#define DTYPE_SHORT  3
#define DTYPE_USHORT 4
#define DTYPE_LONG   5
#define DTYPE_ULONG  6
#define DTYPE_FLOAT  7
#define DTYPE_SBYTE  10
#define DTYPE_LONG2SHORT   15
#define DTYPE_ULONG2SHORT  16

/* input image types */
#define ITYPE_GREY    1
#define ITYPE_COMPLEX 2
#define ITYPE_RGB     3

/* compression */
#define CTYPE_NONE 0
static void rawswap(unsigned char *indata, int pixsize, int xysize);
static void rawdecompress
(void *indata, int pixsize, int xysize, int compression);

static int setintype(char *stype, int *size, int *otype);

void usage(void)
{     
  printf("raw2mrc %s %s %s.\n", VERSION_NAME, __DATE__,
          __TIME__);
  imodCopyright();
  printf("Converts raw data into mrc file format.\n");
  printf("Usage raw2mrc [options] <input files...> <output file>\n");
  printf("Options:\n");
  printf("\t-x #    Width of input image, default 640\n");
  printf("\t-y #    Height of input image, default 480\n");
  printf("\t-z #    Number of sections in input, default 1\n");
  printf("\t-t type Input data type. \n"
          "\t        (byte, sbyte, short, ushort, "
          "long, ulong, float, rgb)\n");
  printf("\t-s      Swap input bytes.\n");
  printf("\t-o #    Offset at beginning of file, default 0\n");
  printf("\t-oz #   Offset between each pair of sections, "
          "default 0\n");
  printf("\t-f      Flip image around X axis\n");
  printf("\t-i      Stack images in inverted order in Z\n");
  printf("\t-d      Divide unsigned shorts by 2 instead of subtracting 32767\n");
  printf("\t-u      Save unsigned shorts in unsigned file mode (6) \n");
  printf("\t-c      Convert long or ulong input to 16-bit"
          " integers, not floats\n");
  printf("\t-p #    Pixel size in Angstroms to set in file header\n");
  printf("\t-pz #   Pixel size in Z in Angstroms, if different\n");
  return;
}

int main( int argc, char *argv[] )
{
  struct MRCheader hdata;
  int    i, j, k, zread, jread, iarg;
  unsigned int xysize;
  FILE   *fin;
  FILE   *fout;
  int    intype = -1;
  int    outtype;
  int    outPixsize = -1;
  int    compression = 0;
  int    byteswap = FALSE;
  int    x = 640, y = 480, z = 1;
  int    nfiles, nsecs;
  int    flip = 0;   /* flag whether to flip about X axis */
  int    divide = 0; /* flag to divide unsigned short by 2 */
  int    keepUshort = 0;  /* Flag to keep unsigned shorts */
  int    hsize = 0;  /* amount of data to skip */
  int    convert = 0; /* Flag to convert long to short */
  int    invertStack = 0;
  int    invertFiles = 0;
  int    zoffset = 0;
  int    pixsize = 1;
  float  pixel;
  double pixSpacing = 0., zPixel = 0.;
  float  mean = 0.0f, tmean = 0.0f, max = -5e29f, min= 5e29f;
  int start = 0;
  b3dByte *sbdata;
  b3dUByte *bdata;
  b3dInt16 *sdata;
  b3dUInt16 *usdata;
  b3dInt32 *ldata;
  b3dUInt32 *uldata;
  b3dFloat *fdata;
  void *indata;
  int val;
  char *progname = imodProgName(argv[0]);
  char prefix[100];
  sprintf(prefix, "ERROR: %s - ", progname);
  setExitPrefix(prefix);

  if (argc < 3){
    usage();
    exit(3);
  }
     

  for (iarg = 1; iarg < argc ; iarg++){
    if (argv[iarg][0] == '-' && strlen(argv[iarg]) <= 3){
      switch (argv[iarg][1]){

      case 't': /* input file type */
        intype = setintype(argv[++iarg], &pixsize, &outtype);
        /* printf("pixsize = %d\n", pixsize); */
        break;
            
      case 'o':
        if (argv[iarg][2] == 'z'){
          zoffset = atoi(argv[++iarg]);
        }else{
          hsize = atoi(argv[++iarg]);
        }
        break;
      case 'x':
        x = atoi(argv[++iarg]);
        break;
      case 'y':
        y = atoi(argv[++iarg]);
        break;
      case 'z':
        z = atoi(argv[++iarg]);
        break;

      case 's':
        byteswap = TRUE;
        break;
      case 'f':
        flip = 1;
        break;
      case 'i':
        if (argv[iarg][2] == 'z') {
          invertStack = 1;
        } else if (argv[iarg][2] == 0x00) {
          invertFiles = 1;
          invertStack = 1;
        } else
          exitError("Unrecognized option %s", argv[iarg]);
        break;
      case 'd':
        divide = 1;
        break;
      case 'u':
        keepUshort = 1;
        break;
      case 'c':
        convert = 1;
        break;
      case 'p':
        if (argv[iarg][2] == 'z') {
          zPixel = atof(argv[++iarg]);
        } else if (argv[iarg][2] == 0x00) {
          pixSpacing = atof(argv[++iarg]);
        } else
          exitError("Unrecognized option %s", argv[iarg]);
        break;
      }
    }else break;
  }
     
  if ( (argc - 1) < (iarg + 1)){
    printf("ERROR: %s - insufficient arguments\n", progname);
    usage();
    exit(3);
  }

  if (convert) {
    if (intype < 0 || intype == DTYPE_LONG) {
      intype = setintype("long2short", &pixsize, &outtype);
      outPixsize = 2;
    } else if (intype == DTYPE_ULONG) {
      intype = setintype("ulong2short", &pixsize, &outtype);
      outPixsize = 2;
    } else
      printf("WARNING: %s - conversion option -c ignored with "
             "specified input type\n", progname);
  }

  if (divide && keepUshort) 
    exitError("You cannot divide by 2 and keep mode as unsigned");

  if (intype < 0)
    intype = setintype("byte", &pixsize, &outtype);
  if (outPixsize < 0)
    outPixsize = pixsize;

  if (keepUshort) {
    if (intype == DTYPE_USHORT || intype == DTYPE_ULONG2SHORT)
      outtype = MRC_MODE_USHORT;
    else
      printf("WARNING: %s - option -u ignored with specified input type\n", progname);
  }

  nfiles = argc - (iarg + 1);
  nsecs = z *nfiles;

  /* printf("nfiles = %d, argc = %d\n",nfiles, argc); */

  if (!getenv("IMOD_NO_IMAGE_BACKUP") && imodBackupFile(argv[argc - 1])) 
    exitError("Couldn't create backup file");
  fout = fopen(argv[argc - 1], "wb");
  if (!fout)
    exitError("Opening %s for output", argv[argc -1]);

  for (j = 0; j < 1024; j++)
    b3dFwrite(&start , 1, 1, fout);

  xysize = x * y;

  indata = (void *)malloc(pixsize * xysize);
  if (!indata)
    exitError("Getting memory");

  /* 7/21/11: make header now so signed byte output is known */
  mrc_head_new(&hdata, x, y, nsecs, outtype);
  if (pixSpacing > 0) {
    if (!zPixel)
      zPixel = pixSpacing;
    mrc_set_scale(&hdata, pixSpacing, pixSpacing, zPixel);
  }

  for (j = iarg ; j < argc-1 ; j++) {
    jread = invertFiles ? iarg + argc - 2 - j : j;
    fin = fopen(argv[jread], "rb");
    if (!fin)
      exitError("Opening %s for input", argv[jread]);

    for (k = 0; k < z; k++) {
      tmean = 0.0f;
      zread = invertStack ? z - 1 - k : k;
      if (mrc_big_seek(fin, hsize + zread * zoffset, pixsize * x, y * zread, SEEK_SET))
        exitError("Seeking to data at section %d in file  %s", zread, argv[jread]);
      if (b3dFread(indata, pixsize, xysize, fin) != xysize) 
        exitError("Reading data from file  %s", argv[jread]);
      if (compression){
        rawdecompress(indata, pixsize, xysize, compression);
      }
      if (byteswap)
        rawswap(indata, pixsize, xysize);


      /* First do any needed conversions of the input types */
      switch (intype) {

      case DTYPE_USHORT:
        usdata = (b3dUInt16 *)indata;
        sdata = (b3dInt16 *)indata;
        if (divide) 
          for (i = 0; i < xysize; i++)
            sdata[i] = (b3dInt16)((int)(usdata[i]) / 2);
        else if (!keepUshort)
          for (i = 0; i < xysize; i++)
            sdata[i] = (b3dInt16)((int)(usdata[i]) - 32767);
        break;

      case DTYPE_LONG:
        ldata = (b3dInt32 *)indata;
        fdata = (b3dFloat *)indata;
        for (i = 0; i < xysize; i++)
          fdata[i] = (b3dFloat)ldata[i];
        break;

      case DTYPE_ULONG:
        uldata = (b3dUInt32 *)indata;
        fdata = (b3dFloat *)indata;
        for (i = 0; i < xysize; i++)
          fdata[i] = (b3dFloat)uldata[i];
        break;

      case DTYPE_ULONG2SHORT:
        uldata = (b3dUInt32 *)indata;
        sdata = (b3dInt16 *)indata;
        usdata = (b3dUInt16 *)indata;
        if (divide) 
          for (i = 0; i < xysize; i++)
            sdata[i] = (b3dInt16)(uldata[i] / 2);
        else if (keepUshort)
          for (i = 0; i < xysize; i++)
            usdata[i] = (b3dUInt16)uldata[i];
        else
          for (i = 0; i < xysize; i++)
            sdata[i] = (b3dInt16)(uldata[i] - 32767);
        break;

      case DTYPE_LONG2SHORT:
        ldata = (b3dInt32 *)indata;
        sdata = (b3dInt16 *)indata;
        if (divide) 
          for (i = 0; i < xysize; i++)
            sdata[i] = (b3dInt16)(ldata[i] / 2);
        else
          for (i = 0; i < xysize; i++)
            sdata[i] = (b3dInt16)ldata[i];
        break;

        /* DNM 12/27/01: fixed these two types on the PC by making the 
           signed and unsigned types explicit */
      case DTYPE_SBYTE:
        sbdata  = (b3dByte *)indata;
        bdata = (b3dUByte *)indata;
        for (i = 0; i < xysize; i++) {
          val = sbdata[i];
          val += 128;
          bdata[i] = val;
        }
        break;

      default:
        break;
      }


      /* Now compute min, mnax, mean based on output type */
      switch (outtype) {
        
      case MRC_MODE_SHORT:
        sdata = (b3dInt16 *)indata;
        for (i = 0; i < xysize; i++) {
          pixel = sdata[i];
          tmean += pixel;
          if (min > pixel)
            min = pixel;
          if (max < pixel)
            max = pixel;
        }
        break;

      case MRC_MODE_USHORT:
        usdata = (b3dUInt16 *)indata;
        for (i = 0; i < xysize; i++) {
          pixel = usdata[i];
          tmean += pixel;
          if (min > pixel)
            min = pixel;
          if (max < pixel)
            max = pixel;
        }
        break;

      case MRC_MODE_FLOAT:
        fdata = (b3dFloat *)indata;
        for (i = 0; i < xysize; i++) {
          pixel = fdata[i];
          tmean += pixel;
          if (min > pixel)
            min = pixel;
          if (max < pixel)
            max = pixel;
        }
        break;

      case MRC_MODE_BYTE:
        bdata = (b3dUByte *)indata;
        for (i = 0; i < xysize; i++) {
          pixel = bdata[i];
          tmean += pixel;
          if (min > pixel)
            min = pixel;
          if (max < pixel)
            max = pixel;
        }
        b3dShiftBytes(bdata, (char *)bdata, x, y, 1, hdata.bytesSigned);
        break;

      default:
        break;
      }

      if (flip == 0) {
        if (b3dFwrite(indata, outPixsize, xysize, fout) != xysize) 
          exitError("Writing data to file");
      } else {

        for (i = y - 1; i >= 0 ; i--) {
          bdata = ((unsigned char *)indata) + i * outPixsize * x;
          if (b3dFwrite(bdata, outPixsize, x, fout) != x) 
            exitError("Writing data to file");
        }
      }
      mean += (tmean / (float)xysize);
    }
    fclose(fin);
  }

  mean  /= (float)nsecs;

  printf("Min = %g, Max = %g, Mean = %g\n", min, max, mean);

  /* write out MRC header */
  /* DNM 11/5/02: change from raw writes of each element to calling
     library routines.  Added label.  1/17/04: eliminate unneeded rewind */
  hdata.amin = min;
  hdata.amax = max;
  hdata.amean = mean;
  mrc_head_label(&hdata, "raw2mrc: Converted to mrc format.");
  mrc_head_write(fout, &hdata);

  fclose(fout);

  exit(0);
}


static void rawdecompress
(void *indata, int pixsize, int xysize, int compression)
{
  if (!compression)
    return;

  return;
}


static void rawswap(b3dUByte *indata, int pixsize, int xysize)
{
  b3dUByte u1,u2,u3,u4;
  int i;

  if (pixsize == 4) {
    for (i = 0 ; i < xysize; i++) {
           
      /* DNM 12/18/01: changed i * 2 to i * 4 */
      u1 = indata[i * 4];
      u2 = indata[(i * 4) + 1];
      u3 = indata[(i * 4) + 2];
      u4 = indata[(i * 4) + 3];
      indata[i * 4] = u4;
      indata[(i * 4) + 1] = u3;
      indata[(i * 4) + 2] = u2;
      indata[(i * 4) + 3] = u1;
    }
    return;
  }


  if (pixsize == 2) {
    for (i = 0 ; i < xysize; i++) {
      u1 = indata[i * 2];
      u2 = indata[(i * 2) + 1];
      indata[i * 2] = u2;
      indata[(i * 2) + 1] = u1;
    }
    return;
  }
  return;
}

static int setintype(char *stype, int *size, int *otype)
{
  if (!strcmp(stype,"byte")) {
    *size = 1;
    *otype = MRC_MODE_BYTE;
    return(DTYPE_BYTE);
  }
     
  if (!strcmp(stype,"sbyte")) {
    *size = 1;
    *otype = MRC_MODE_BYTE;
    return(DTYPE_SBYTE);
  }

  if (!strcmp(stype,"rgb")) {
    *size = 3;
    *otype = MRC_MODE_RGB;
    return(DTYPE_BYTE);
  }

  if (!strcmp(stype, "short")) {
    *size = 2;
    *otype = MRC_MODE_SHORT;
    return(DTYPE_SHORT);
  }
  if (!strcmp(stype,"ushort")) {
    *size = 2;
    *otype = MRC_MODE_SHORT;
    return(DTYPE_USHORT);
  }

  /* DNM 12/27/01: long was missing from the quotes here */
  if (!strcmp(stype,"long")) {
    *size = 4;
    *otype = MRC_MODE_FLOAT;
    return(DTYPE_LONG);
  }
  if (!strcmp(stype,"ulong")) {
    *size = 4;
    *otype = MRC_MODE_FLOAT;
    return(DTYPE_ULONG);
  }

  if (!strcmp(stype,"long2short")) {
    *size = 4;
    *otype = MRC_MODE_SHORT;
    return(DTYPE_LONG2SHORT);
  }
  if (!strcmp(stype,"ulong2short")) {
    *size = 4;
    *otype = MRC_MODE_SHORT;
    return(DTYPE_ULONG2SHORT);
  }
     
  if (!strcmp(stype,"float")) {
    *size = 4;
    *otype = MRC_MODE_FLOAT;
    return(DTYPE_FLOAT);
  }

  return(-1);
}
